File: Animations/Linked Particles/TypeScript/src/particles.ts
import { Mouse } from "./mouse"

export class Particle {
    private x!: number
    private y!: number
    private radius = Math.floor((Math.random() * 15) + 1)
    private dx = (Math.random() * 2) - 1
    private dy = (Math.random() * 2) - 1
    private pushForceX = 0
    private pushForceY = 0
    private friction: number

    constructor(
        private context: CanvasRenderingContext2D | null,
        private width: number,
        private height: number,
        private mouse: Mouse,
    ) {
        this.setPosition()
        this.friction = this.calculateFriction()
    }

    // --------------------------------------------------------------------------------------------

    // Keeps particles within the canvas. 

    private bounds(): void {
        if (this.x < 0 + this.radius) {
            this.x = this.radius
            this.dx *= -1
        }
        if (this.x > this.width - this.radius) {
            this.x = this.width - this.radius
            this.dx *= -1
        }

        if (this.y < 0 + this.radius) {
            this.y = this.radius
            this.dy *= -1
        }
        if (this.y > this.height - this.radius) {
            this.y = this.height - this.radius
            this.dy *= - 1
        }
    }

    // --------------------------------------------------------------------------------------------

    // Set a friction value base on size of particle.

    private calculateFriction(): number {
        if (this.radius <= 7) {
            return 0.99
        } else if (this.radius <= 10) {
            return 0.975
        } else {
            return 0.95
        }
    }

    // --------------------------------------------------------------------------------------------

    // Draw the individual particle on the screen.

    public draw(): void {
        if (this.context != null) {
            this.context.beginPath()
            this.context.arc(this.x, this.y, this.radius, 0, 2 * Math.PI, false)
            //this.context.fillStyle = `hsl(${this.x}, 100%, 50%)`
            this.context.fill()
            this.context.stroke()
        }
    }

    // --------------------------------------------------------------------------------------------

    // Returns the X and Y values of the particle.

    public getPosition(): number[] {
        return [this.x, this.y]
    }

    // --------------------------------------------------------------------------------------------

    // Returns the radius of the particle.

    public getRadius(): number {
        return this.radius
    }

    // --------------------------------------------------------------------------------------------

    // Changers the particle coordinates. 

    private move(): void {
        this.x += this.dx + this.pushForceX
        this.y += this.dy + this.pushForceY
    }

    // --------------------------------------------------------------------------------------------

    private pushWithMouse(): void {
        if (this.mouse.getPressedStatus()) {
            let mousePositions = this.mouse.getPosition()
            let dx = this.x - mousePositions[0]
            let dy = this.y - mousePositions[1]
            let distance = Math.hypot(dx, dy)

            if (distance < this.mouse.getRadius()) {
                let force = this.mouse.getRadius() / distance
                let angle = Math.atan2(dy, dx)
                this.pushForceX = Math.cos(angle) * force
                this.pushForceY = Math.sin(angle) * force
            }
        }
    }

    // --------------------------------------------------------------------------------------------

    // Changes width and height variables and resets the particle postion.

    public resizeCanvas(width: number, height: number): void {
        this.width = width
        this.height = height
        this.setPosition()
    }


    // --------------------------------------------------------------------------------------------

    // Set the x and y values of the particles inside the limits of the canvas.

    private setPosition(): void {
        this.x = (Math.random() * this.width - (this.radius * 2)) + this.radius * 2
        this.y = (Math.random() * this.height - (this.radius * 2)) + this.radius * 2
    }

    // --------------------------------------------------------------------------------------------

    // Update particle state.

    public update(): void {
        this.pushWithMouse()
        this.move()
        this.bounds()
        this.updateForce()
    }

    // --------------------------------------------------------------------------------------------

    // Adds friction to the force applied to particles.

    private updateForce(): void {
        this.pushForceX *= this.friction
        this.pushForceY *= this.friction
    }

    // --------------------------------------------------------------------------------------------

}

File: Animations/Linked Particles/Go/particles.go
package main

import (
	"math"
	"math/rand"

	rl "github.com/gen2brain/raylib-go/raylib"
)

type Particle struct {
	width               int32
	height              int32
	radius              int32
	x                   int32
	y                   int32
	movementX           int32
	movementY           int32
	powerPushMultiplier float64
	pushX               float64
	pushY               float64
	friction            float64
	mouse               *Mouse
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) init() {
	p.setMovementSpeed()
	p.radius = int32(rand.Intn(10)) + 5
	p.pushX = 0
	p.pushY = 0
	p.setFriction()
	p.setPosition()
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) collision() {
	if p.radius > p.x {
		p.x = 1 + p.radius
		p.movementX *= -1
		p.pushX *= -1
	}
	if p.x+p.radius > p.width {
		p.x = p.width - p.radius - 1
		p.movementX *= -1
		p.pushX *= -1
	}

	if p.radius > p.y {
		p.y = 1 + p.radius
		p.movementY *= -1
		p.pushY *= -1
	}
	if p.y+p.radius > p.height {
		p.y = p.height - p.radius - 1
		p.movementY *= -1
		p.pushY *= -1
	}
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) draw() {
	color := rl.ColorFromHSV(float32(p.x), 1.0, 1.0)
	rl.DrawCircle(p.x, p.y, float32(p.radius), color)
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) movement() {
	p.x += p.movementX + int32(math.Round(p.pushX))
	p.y += p.movementY + int32(math.Round(p.pushY))
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) pushParticle() {
	mx, my := p.mouse.getLocation()
	dx := float64(p.x - *mx)
	dy := float64(p.y - *my)

	distance := math.Hypot(dx, dy)
	if distance < float64(*p.mouse.getRadius()) {
		power := (float64(*p.mouse.getRadius()) / distance) * p.powerPushMultiplier
		angle := math.Atan2(dy, dx)
		p.pushX = math.Cos(angle) * power
		p.pushY = math.Sin(angle) * power
	}
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) setFriction() {
	if p.radius <= 7 {
		p.friction = 0.99
	} else if p.radius > 7 && p.radius <= 10 {
		p.friction = 0.97
	} else {
		p.friction = 0.94
	}
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) setMovementSpeed() {
	p.movementX = int32(rand.Intn(4)) - 2
	p.movementY = int32(rand.Intn(4)) - 2
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) setPosition() {
	p.x = int32(rand.Intn(int(p.width)-int(p.radius))) + p.radius
	p.y = int32(rand.Intn(int(p.height)-int(p.radius))) + p.radius
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) update() {
	if *p.mouse.activationStatus() {
		p.pushParticle()
	}
	p.movement()
	p.collision()
	p.updatePush()
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) updatePush() {
	p.pushX *= p.friction
	p.pushY *= p.friction
}

// ---------------------------------------------------------------------------------------------------------------------

func (p *Particle) updateWindowSize() {
	p.width = int32(rl.GetScreenWidth())
	p.height = int32(rl.GetScreenHeight())
}

// ---------------------------------------------------------------------------------------------------------------------
